runtime/recipe/type-checker.js

```diff
@@ -97,18 +97,24 @@ class TypeChecker {
     } else {
       assert(false, 'tryMergeTypeVariable shouldn\'t be called on two types without any type variables');
     }
-    
+
     return base;
   }
 
   static _tryMergeConstraints(handleType, {type, direction}) {
     let [primitiveHandleType, primitiveConnectionType] = Type.unwrapPair(handleType.resolvedType(), type.resolvedType());
     if (primitiveHandleType.isVariable) {
-      // if this is an undifferentiated variable then we need to create structure to match against. That's
-      // allowed because this variable could represent anything, and it needs to represent this structure
-      // in order for type resolution to succeed.
       if (primitiveConnectionType.isSetView) {
-        assert(primitiveHandleType.variable.resolution == null && primitiveHandleType.variable.canReadSubset == null && primitiveHandleType.variable.canWriteSuperset == null);
+        if (primitiveHandleType.variable.resolution != null
+            || primitiveHandleType.variable.canReadSubset != null
+            || primitiveHandleType.variable.canWriteSuperset != null) {
+          // If this is a variable with existing resolution or constraints,
+          // we cannot merge SetView constraints.
```

Comment:
maybe "resolved / constrained variables can only represent individual Entities, not sets"?